### PR TITLE
[Feature] Support openset kie

### PR DESCRIPTION
**Pull request by cuhk-hbsun**

### CODE REVIEW

**Comment by amitbcp** [comment]
@cuhk-hbsun since this PR is in review for a long time, can you please share the OpenSet data via the docs or MMOCR website, I would like to start preparing my data and understand the new labels requirement till then. Thanks in advance.

**Comment by INF800** [comment]
Hi @gaotongxiao. How is this different from wild receipt dataset?

**Comment by cuhk-hbsun** [comment]
@amitbcp The annotation files are attache
[openset_test.txt](https://github.com/open-mmlab/mmocr/files/7309376/openset_test.txt)
[openset_train.txt](https://github.com/open-mmlab/mmocr/files/7309378/openset_train.txt)
d.

**Comment by gaotongxiao** [comment]
@INF800 It adds everything that you would expect - model, dataset, docs, training and inference code. Please excuse us for the long review process :)

**Comment by amitbcp** [comment]
@cuhk-hbsun just to confirm, so now while we train, the edge loss predicts whether the edge should exist or not based on the labels ?

Also, suppose we have words which should not be connected to any other words (like background words) then the edge label will be unique for such a background word ?


lastly, if we want to connect more than one words then should they have the same edge label ?
Example1  Text : "My name is Amit Agarwal."
Doubt : Since this is a sentence or a single entity, can the edge label for all the 5 words be same ?

Example 2. Text : "My name is Amit Kumar Agarwal. "
Doubt : The word Amit and Agarwal are connected. (first name ,last name). Also the word Amit and Kumar are connected. (first name, middle name).

How to handle this in the annotation ? or this is currently not supported ?

**Comment by cuhk-hbsun** [comment]
> @cuhk-hbsun just to confirm, so now while we train, the edge loss predicts whether the edge should exist or not based on the labels ?  （Yes）
> 
> Also, suppose we have words which should not be connected to any other words (like background words) then the edge label will be unique for such a background word ?  （Yes）
> 
> lastly, if we want to connect more than one words then should they have the same edge label ? Example1 Text : "My name is Amit Agarwal." Doubt : Since this is a sentence or a single entity, can the edge label for all the 5 words be same ? (Maybe No)
> 
> Example 2. Text : "My name is Amit Kumar Agarwal. " Doubt : The word Amit and Agarwal are connected. (first name ,last name). Also the word Amit and Kumar are connected. (first name, middle name). 
> 
> How to handle this in the annotation ? or this is currently not supported ?


The valid edge is just for key & value pair.

For example, there are `hello`, `world`, `actor`, `James`, `Tom`, `something`, `actress`, `Lucy`, `Zora` in one image. The labels should be as follows:
|box_content | closeset_node_label| closeset_edge_label | openset_node_label | openset_edge_label |
| :----: | :---: | :----: | :---: | :---: |
| hello | 0 | - | 0 | 0 |
| world | 0 | - | 0 | 1 |
| Actor | 1 | - | 1 | 2 |
| Tom | 2 | - | 2 | 2 |
| Tony | 2 | - | 2 | 2 |
| Tim | 2 | - | 2 | 2 |
| something | 0 | - | 0 | 3 |
| Actress | 3 | - | 1 | 4 |
| Lucy | 4 | - | 2 | 4 |
| Zora | 4 | - | 2 | 4 |

For `openset` annotation, the `key` and `value` with the same edge label will have a valid edge, like `Actor` and `Tom`. But `Tom` and `Tim` have no valid edge. 
In other words, there are no valid edges among `values` in the `openset wildreceipt`. But you can have a try with a different annotation format.

**Comment by amitbcp** [comment]
> @cuhk-hbsun thanks for the explanation. I still have a few questions based on your previous explanation :
> 
> 1. `For openset annotation, the key and value with the same edge label will have a valid edge` : I believe this should be configurable in the project so that as anyone from the community uses it, they can define which set of labels/words can have valid edges.
> 2. I hope with the openest annotation we are restricting the number of labels right ? Like I can have still use a custom data in the proposed openset format to have more the 4 label classes ?
> 3. `In other words, there are no valid edges among values in the openset wildreceipt. But you can have a try with a different annotation format.` : Can you please elaborate how we can try different annotation format that is supported in the framework ?
> 4. Also, just wanted to understand the importance of `closeset_node_label` in this process, because the wildreciept and the paper I could not find any reference all importance of closest node

@cuhk-hbsun  can you please have a look at the followup questions ?

**Comment by cuhk-hbsun** [comment]
> > @cuhk-hbsun thanks for the explanation. I still have a few questions based on your previous explanation :
> > 
> > 1. `For openset annotation, the key and value with the same edge label will have a valid edge` : I believe this should be configurable in the project so that as anyone from the community uses it, they can define which set of labels/words can have valid edges.
> > 2. I hope with the openest annotation we are restricting the number of labels right ? Like I can have still use a custom data in the proposed openset format to have more the 4 label classes ?
> > 3. `In other words, there are no valid edges among values in the openset wildreceipt. But you can have a try with a different annotation format.` : Can you please elaborate how we can try different annotation format that is supported in the framework ?
> > 4. Also, just wanted to understand the importance of `closeset_node_label` in this process, because the wildreciept and the paper I could not find any reference all importance of closest node
> 
> @cuhk-hbsun can you please have a look at the followup questions ?

1. Yes, just have a try and see if it works.
2. Yes, you can set the number_classes of the node. But some functions maybe not work like `evaluete` in `openset_kie_dataset.py` and `imshow_edge` in `visulaize.py`.
3. In `openset_kie_dataset.py`, for nodes with the same edge index, only node with key label and node with value label will give an valid edge. Edge among values is not supported now.
4. In Table 3 of the paper, the categories are the node classes.  For example, `store_name_key` and `store_name_value` are two types of nodes.

File: tools/data/kie/closeset_to_openset.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            if edge is not None:
+                anno['edge'] = edge
+                if label in closeset_key_inds:
+                    anno['label'] = openset_node_label_mapping['key']
+                elif label in closeset_value_inds:
+                    anno['label'] = openset_node_label_mapping['value']
+            else:
+                tmp_key = 'key'
+                if label in closeset_key_inds:
+                    label_with_same_edge = label - 1
```
Comment (gaotongxiao) [inline tools/data/kie/closeset_to_openset.py:68]:
```suggestion
                    label_with_same_edge = closeset_value_inds[
                        closeset_key_inds.index(label)]
```
Could make it more general if the assumption in L21-24 holds true

File: tools/data/kie/closeset_to_openset.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                elif label in closeset_value_inds:
+                    label_with_same_edge = label + 1
```
Comment (gaotongxiao) [inline tools/data/kie/closeset_to_openset.py:70]:
```suggestion
                    label_with_same_edge = closeset_key_inds[
                        closeset_value_inds.index(label)]
```

**Comment by amitbcp** [comment]
Thanks @cuhk-hbsun  for explaining. I'll check it out.

I am still having a confusion on this ```Yes, you can set the number_classes of the node. But some functions maybe not work like evaluete in openset_kie_dataset.py and imshow_edge in visulaize.py.```

If I increase the number_classes of the nodes, how to I still state valid edges ? For example : key class and value class have a valid edge by default. If I add 2 more labels ("store_name_key" and "store_name_value") how do I specify edges between these two labels in a custom dataset ?

**Review by gaotongxiao — COMMENTED** [review]
LGTM, I'll supplement some docs for the final approval

**Comment by cuhk-hbsun** [comment]
> Thanks @cuhk-hbsun for explaining. I'll check it out.
> 
> I am still having a confusion on this `Yes, you can set the number_classes of the node. But some functions maybe not work like evaluete in openset_kie_dataset.py and imshow_edge in visulaize.py.`
> 
> If I increase the number_classes of the nodes, how to I still state valid edges ? For example : key class and value class have a valid edge by default. If I add 2 more labels ("store_name_key" and "store_name_value") how do I specify edges between these two labels in a custom dataset ?

For training, it's fine to add 2 more labels directly.
For testing, [deocder_pairs_gt](https://github.com/cuhk-hbsun/mmocr-2/blob/hbsun/openset_kie/mmocr/datasets/openset_kie_dataset.py#L103) should be modified to prepare gt pairs as well as predicted pairs.

**Review by gaotongxiao — APPROVED** [review]
